### PR TITLE
Chapter 14: Backtest statistics

### PR DESCRIPTION
**Pull request by PanPip**
Add Chapter 14 code (Timing of flattening and flips, average holding period, bets concentration, drawdown and time under water, Sharpe ratio, probabalistic Sharpe ratio, deflated Sharpe ratio).

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Each function was tested with a self-generated series of data. Formulas for probabalistic Sharpe ratio and deflated Sharpe ratio were tested using numerical examples from book and papers by Lopez de Prado. 

**Test Configuration**:
* Windows 10
* PyCharm

Test file:
- test_backtest_statistics.py

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: mlfinlab/backtest_statistics/__init__.py [hunk 1]
```diff
@@ -0,0 +1,7 @@
+"""
+Functions derived from Chapter 14: Backtest Statistics
```
Comment (Jackal08) [inline mlfinlab/backtest_statistics/__init__.py]: Let's use a better description. Explain the module and what code is included in it. I want to stress the depth I desire and how important these docstrings are. They are almost more important than the code written. They form the basis for how users will make use of your code.

File: mlfinlab/backtest_statistics/statistics.py [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+    :param target_positions: (pd.Series) target position series with timestamps as indices
+    :return: (pd.DatetimeIndex) timestamps of trades flattening, flipping and last bet
+    """
+    empty_positions = target_positions[(target_positions==0)].index # Empty positions index
+    previous_positions = target_positions.shift(1) # Timestamps pointing at previous positions
+    previous_positions = previous_positions[(previous_positions!=0)].index # Index of positions where previous one wasn't empty
+    flattening = empty_positions.intersection(previous_positions) # FLATTENING - if previous position was open, but current is empty
+    multiplied_posions = target_positions.iloc[1:] * target_positions.iloc[:-1].values # Multiplies current position with value of next one
+    flips = multiplied_posions[(multiplied_posions<0)].index # FLIPS - if current position has another direction compared to the next
+    res = flattening.union(flips).sort_values()
```
Comment (Jackal08) [inline mlfinlab/backtest_statistics/statistics.py]: Better variable names, please.

File: mlfinlab/backtest_statistics/statistics.py [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+    - entry_time = (previous_time * weight_of_previous_position + 
+                    time_since_beginning_of_trade * increase_in_position ) /
+                    weight_of_current_position
+    - holding_period ['holding_time' = time a position was held, 'weight' = weight of position closed]
+    - res = weighted average time a trade was held
+    
+    :param target_positions: (pd.Series) target position series with timestamps as indices
+    :return: (float) estimated average holding period, NaN if zero or unpredicted
+    """
+    holding_period, entry_time = pd.DataFrame(columns=['holding_time','weight']), 0
```
Comment (Jackal08) [inline mlfinlab/backtest_statistics/statistics.py]: Avoid stacking variables on multiple lines. entry_time on its own line, please.

File: mlfinlab/backtest_statistics/statistics.py [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+    position_difference = target_positions.diff() 
+    
+    # Time elapsed from the starting time for each position
+    time_difference = (target_positions.index-target_positions.index[0]) / np.timedelta64(1,'D')
+    for i in range(1, target_positions.shape[0]):
+        if float(position_difference.iloc[i] * target_positions.iloc[i-1]) >= 0: # Increased or unchanged position
+            if float(target_positions.iloc[i]) != 0: # And not an empty position
+                entry_time = (entry_time * target_positions.iloc[i-1] + 
+                              time_difference[i] * position_difference.iloc[i]) / target_positions.iloc[i]
+        else: # Decreased
```
Comment (Jackal08) [inline mlfinlab/backtest_statistics/statistics.py]: I like the use of inline comments here.

File: mlfinlab/backtest_statistics/statistics.py [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+            if float(target_positions.iloc[i] * target_positions.iloc[i-1]) < 0: # Flip of a position
```
Comment (Jackal08) [inline mlfinlab/backtest_statistics/statistics.py]: Code is a bit long. What is your line length set to? Lets reduce it so that all the code will fit on Github review with out the need to scroll.

File: mlfinlab/backtest_statistics/statistics.py [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+                holding_period.loc[target_positions.index[i], ['holding_time','weight']] = (time_difference[i] - entry_time, abs(target_positions.iloc[i-1]))
+                entry_time = time_difference[i] # Reset entry time
+            else: # Only a part of position is closed
+                holding_period.loc[target_positions.index[i], ['holding_time','weight']] = (time_difference[i] - entry_time, abs(position_difference.iloc[i]))
+                
+    if float(holding_period['weight'].sum()) > 0: # If there were closed trades at all
+        res = float((holding_period['holding_time'] * holding_period['weight']).sum() / holding_period['weight'].sum())
```
Comment (Jackal08) [inline mlfinlab/backtest_statistics/statistics.py]: variable names. ```res```

File: mlfinlab/backtest_statistics/statistics.py [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+    """  
+    if returns.shape[0]<=2:
+        return float('nan') # If less than 3 bets
+    weights = returns / returns.sum() # Weights of each bet
+    hhi = (weights ** 2).sum() # Herfindahl-Hirschman Index for weights
+    hhi = float((hhi - returns.shape[0] **(-1)) / (1 - returns.shape[0]**(-1)))
+    return hhi
+
+
+def positive_negative_bets_concentration(returns: pd.Series, freq: str = 'M') -> tuple:
```
Comment (Jackal08) [inline mlfinlab/backtest_statistics/statistics.py]: @proskurin, I have mixed feelings about using the strong typed variables here. I think it makes the documentation on ReadTheDocs less readable. I can understand why its desired from a code point of view but these types are also not enforced are they?

I'd love to hear your thoughts on this.

Comment (Jackal08) [inline mlfinlab/backtest_statistics/statistics.py]: Better function name?

File: mlfinlab/backtest_statistics/statistics.py [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+    Calculates Annualized Sharpe Ratio for pd.Series of returns.
+        
+    :param returns: (pd.Series) returns
+    :param cumulative: (bool) flag if returns are cumulative (no by default)
+    :param entries_per_year: (int) times returns are recorded per year (days by default)
+    :param risk_free_rate: (float) risk-free rate (0 by default)
+    :return: (float) Annualized Sharpe Ratio
+    """  
+    if cumulative:
+        returns = returns.diff() # Converting to simple returns
```
Comment (Jackal08) [inline mlfinlab/backtest_statistics/statistics.py]: Is diff correct? I think it assumes your looking at the log cumulative returns. Since log is additive. This wont work if they used the normal cumulative returns. Is this how it is in the book?

File: mlfinlab/backtest_statistics/statistics.py [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+    :param observed_SR: (float) Sharpe Ratio that is observed
+    :param benchmark_SR: (float) Sharpe Ratio to which observed_SR is tested against
+    :param  number_of_returns: (int) times returns are recorded for observed_SR
+    :param skewness_of_returns: (float) skewness of returns (as Gaussian by default)
+    :param kurtosis_of_returns: (float) kurtosis of returns (as Gaussian by default)
+    :return: (float) Probabalistic Sharpe Ratio
+    """  
+    
+    
+    PSR = ss.norm.cdf(((observed_SR - benchmark_SR) * (number_of_returns - 1)**(1/2)) / \
```
Comment (Jackal08) [inline mlfinlab/backtest_statistics/statistics.py]: Why capital PSR? That is reserved for class names.

File: mlfinlab/backtest_statistics/statistics.py [hunk 10]
```diff
@@ -0,0 +1,233 @@
...
+    :param skewness_of_returns: (float) skewness of returns (as Gaussian by default)
+    :param kurtosis_of_returns: (float) kurtosis of returns (as Gaussian by default)
+    :return: (float) Deflated Sharpe Ratio
+    """  
+    # Calculating benchmark_SR from a list of estimates
+    benchmark_SR = np.array(SR_estimates).std() * ((1 - np.euler_gamma) * 
+                                                   ss.norm.ppf(1 - 1/len(SR_estimates)) + 
+                                                   np.euler_gamma * ss.norm.ppf(1 - 1/len(SR_estimates) * np.e**(-1)))
+    
+    DSR = probabalistic_sharpe_ratio(observed_SR, benchmark_SR, number_of_returns, skewness_of_returns,
```
Comment (Jackal08) [inline mlfinlab/backtest_statistics/statistics.py]: Same here w.r.t capital letters.

File: mlfinlab/backtest_statistics/statistics.py [hunk 11]
```diff
@@ -0,0 +1,233 @@
+"""
+Implements the book Chapter 14 on Backtest Statistics
```
Comment (Jackal08) [inline mlfinlab/backtest_statistics/statistics.py]: Better docstring please.

File: mlfinlab/backtest_statistics/__init__.py (hunk 1 above)
Comment (PanPip) [reply mlfinlab/backtest_statistics/__init__.py]: Will a list of functions and their short descriptions (as comments in statistics.py) be enough? Or a shorter description (a sentence/function) is preferred?

File: mlfinlab/backtest_statistics/statistics.py (hunk 2 above)
Comment (PanPip) [reply mlfinlab/backtest_statistics/statistics.py]: I'll change _res_ to flips_and_flattening, as it was just a group of both. I was a bit puzzled as haven't found Lopez de Prado mentioning flattening in other parts of the book.

File: mlfinlab/backtest_statistics/statistics.py (hunk 3 above)
Comment (PanPip) [reply mlfinlab/backtest_statistics/statistics.py]: Will change

File: mlfinlab/backtest_statistics/statistics.py (hunk 6 above)
Comment (PanPip) [reply mlfinlab/backtest_statistics/statistics.py]: Will change

File: mlfinlab/backtest_statistics/statistics.py (hunk 7 above)
Comment (PanPip) [reply mlfinlab/backtest_statistics/statistics.py]: Will _specific_bets_concentration_ work?

File: mlfinlab/backtest_statistics/statistics.py (hunk 8 above)
Comment (PanPip) [reply mlfinlab/backtest_statistics/statistics.py]: I was thinking of log returns but didn't state it. My bad. Will make another flag and reversion from normal cumulative returns method. 

The book doesn't specify code for neither of the Sharpe ratios :(

File: mlfinlab/backtest_statistics/statistics.py (hunk 9 above)
Comment (PanPip) [reply mlfinlab/backtest_statistics/statistics.py]: Will change

File: mlfinlab/backtest_statistics/statistics.py (hunk 10 above)
Comment (PanPip) [reply mlfinlab/backtest_statistics/statistics.py]: Will change

File: mlfinlab/backtest_statistics/statistics.py (hunk 11 above)
Comment (PanPip) [reply mlfinlab/backtest_statistics/statistics.py]: Will make same description as in the ___init___.py

File: mlfinlab/backtest_statistics/statistics.py (hunk 5 above)
Comment (PanPip) [reply mlfinlab/backtest_statistics/statistics.py]: I didn't have it specified. Will try to fit in 80-100 characters.

File: mlfinlab/backtest_statistics/__init__.py (hunk 1 above)
Comment (Jackal08) [reply mlfinlab/backtest_statistics/__init__.py]: Let's have the focus be on what this module (file) is about. Eventually, this package will include authors other than de Prado, thus a module should have a purpose. This docstring should highlight that purpose. 

I wouldn't do a list of functions because then we need to update the docstring as we add new functions.

File: mlfinlab/backtest_statistics/statistics.py (hunk 11 above)
Comment (Jackal08) [reply mlfinlab/backtest_statistics/statistics.py]: No, the init file is for the high level folder which can have many files. This file is an individual and we should have a docstring that describes this file.

File: mlfinlab/backtest_statistics/statistics.py (hunk 7 above)
Comment (Jackal08) [reply mlfinlab/backtest_statistics/statistics.py]: distribution_bets_concentration?

Comment (Jackal08) [reply mlfinlab/backtest_statistics/statistics.py]: all_bets_concentration?

File: mlfinlab/backtest_statistics/statistics.py (hunk 8 above)
Comment (Jackal08) [reply mlfinlab/backtest_statistics/statistics.py]: I think the standard is to calculate the sharpe in normal returns and not log returns.

Comment (Jackal08) [reply mlfinlab/backtest_statistics/statistics.py]: why to the power of 0.5? 

```**(1/2)```

Comment (Jackal08) [reply mlfinlab/backtest_statistics/statistics.py]: Usually: 

```
sharpe = returns.mean() / returns.std() * np.sqrt(days_in_year)
```

Oh thats why you take the power of 0.5:

Comment (PanPip) [reply mlfinlab/backtest_statistics/statistics.py]: I've cahnged it to normal and not log returns, left the conversion from cumulative normal returns to normal returns.

File: mlfinlab/backtest_statistics/statistics.py [hunk 12]
```diff
@@ -0,0 +1,268 @@
...
+        drawdown = high_watermarks['hwm'] - high_watermarks['min']
+    else:
+        drawdown = 1 - high_watermarks['min'] / high_watermarks['hwm']
+    time_under_water = ((high_watermarks.index[1:] -
+                         high_watermarks.index[:-1]) / np.timedelta64(1, 'Y')).values
+    time_under_water = pd.Series(time_under_water, index=high_watermarks.index[:-1])
+    return drawdown, time_under_water
+
+
+def sharpe_ratio(returns: pd.Series, cumulative: bool = False,
```
Comment (Jackal08) [inline mlfinlab/backtest_statistics/statistics.py]: This is ok that the signature is on 2 lines because its longer.

File: mlfinlab/backtest_statistics/statistics.py [hunk 13]
```diff
@@ -0,0 +1,268 @@
...
+                               kurtosis_of_returns: float = 3) -> float:
+    """
+    Calculates the probabilistic Sharpe ratio (PSR) that provides an adjusted estimate of SR,
+    by removing the inflationary effect caused by short series with skewed and/or
+    fat-tailed returns.
+    Given a user-defined benchmark Sharpe ratio and an observed Sharpe ratio,
+    PSR estimates the probability that SR ̂is greater than a hypothetical SR.
+    - It should exceed 0.95, for the standard significance level of 5%.
+    - It can be computed on absolute or relative returns.
+    :param observed_sr: (float) Sharpe Ratio that is observed
```
Comment (Jackal08) [inline mlfinlab/backtest_statistics/statistics.py]: White space

File: mlfinlab/tests/test_backtest_statistics.py [hunk 14]
```diff
@@ -0,0 +1,202 @@
+"""
```
Comment (Jackal08) [inline mlfinlab/tests/test_backtest_statistics.py]: Are we also testing for edge cases?

File: mlfinlab/tests/test_backtest_statistics.py [hunk 15]
```diff
@@ -0,0 +1,202 @@
...
+        normal_ret = np.array([0.01, 0.03, 0.02, 0.01, -0.01, 0.02, 0.01, 0.0, -0.01, 0.01])
+        cumulated_ret = np.cumprod(1 + normal_ret)
+
+        self.flip_flattening_positions = pd.Series(data=flip_positions,
+                                                   index=dates)
+        self.flips = pd.DatetimeIndex([dt.datetime(2000, 1, 7)])
+        self.flattenings = pd.DatetimeIndex([dt.datetime(2000, 1, 4), dt.datetime(2000, 1, 10)])
+        self.hold_positions = pd.Series(data=hold_positions, index=dates)
+        self.no_closed_positions = pd.Series(data=no_closed_positions, index=dates)
+        self.negative_positions = pd.Series(data=negative_concentrated,
```
Comment (Jackal08) [inline mlfinlab/tests/test_backtest_statistics.py]: One line please.

File: mlfinlab/backtest_statistics/statistics.py [hunk 16]
```diff
@@ -0,0 +1,268 @@
...
+"""
+
+import pandas as pd
+import scipy.stats as ss
+import numpy as np
+
+
+def timing_of_flattening_and_flips(target_positions: pd.Series) -> pd.DatetimeIndex:
+    """
+    Snippet 14.1, page 197, Derives the timestamps of flattening or flipping
```
[the remark this answers is not in the document]
Comment (PanPip) [reply mlfinlab/backtest_statistics/statistics.py]: Ok, the trick is, in the corresponding book I found no mentions of those. I'll describe them the best I can in docstrings

File: mlfinlab/backtest_statistics/statistics.py [hunk 17]
```diff
@@ -0,0 +1,268 @@
...
+    Snippet 14.2, page 197, Estimates the average holding period (in days) of a strategy,
+    given a pandas series of target positions using average entry time pairing algorithm.
+    Idea of an algorithm:
+    - entry_time = (previous_time * weight_of_previous_position +
+                    time_since_beginning_of_trade * increase_in_position ) /
+                    weight_of_current_position
+    - holding_period ['holding_time' = time a position was held,
+                      'weight' = weight of position closed]
+    - res = weighted average time a trade was held
+    :param target_positions: (pd.Series) target position series with timestamps as indices
```
[the remark this answers is not in the document]
Comment (PanPip) [reply mlfinlab/backtest_statistics/statistics.py]: I'd love to. However it seems to be not according to pylint checks. I've asked @proskurin about that one in particular. Should I still add white space?

File: mlfinlab/backtest_statistics/statistics.py [hunk 18]
```diff
@@ -0,0 +1,268 @@
...
+                      np.timedelta64(1, 'D')
+    for i in range(1, target_positions.shape[0]):
+
+        # Increased or unchanged position
+        if float(position_difference.iloc[i] * target_positions.iloc[i - 1]) >= 0:
+            if float(target_positions.iloc[i]) != 0:  # And not an empty position
+                entry_time = (entry_time * target_positions.iloc[i - 1] +
+                              time_difference[i] * position_difference.iloc[i]) / \
+                             target_positions.iloc[i]
+        else:  # Decreased
```
[the remark this answers is not in the document]
Comment (PanPip) [reply mlfinlab/backtest_statistics/statistics.py]: Sure, will do

File: mlfinlab/backtest_statistics/statistics.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (Jackal08) [reply mlfinlab/backtest_statistics/statistics.py]: hmm ok leave it for now, when you are done with the review I will make the changes and push back to this branch.

[the remark this answers is not in the document]
Comment (PanPip) [reply mlfinlab/backtest_statistics/statistics.py]: Ok, I guess it was a warning of my plugin and not pylint

File: mlfinlab/backtest_statistics/statistics.py [hunk 19]
```diff
@@ -0,0 +1,268 @@
...
+            # Flip of a position
+            if float(target_positions.iloc[i] * target_positions.iloc[i - 1]) < 0:
+                holding_period.loc[target_positions.index[i],
```
[the remark this answers is not in the document]
Comment (PanPip) [reply mlfinlab/backtest_statistics/statistics.py]: Will do

File: mlfinlab/tests/test_backtest_statistics.py (hunk 14 above)
Comment (PanPip) [reply mlfinlab/tests/test_backtest_statistics.py]: I can add more examples. By edge cases you mean something like nan values or just higher range of test values?

Comment (Jackal08) [reply mlfinlab/tests/test_backtest_statistics.py]: Yea testing for things like nans or any other strange values that could be passed to the data. 

Does the function opperate as expected under edge case examples.

File: mlfinlab/backtest_statistics/statistics.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (PanPip) [reply mlfinlab/backtest_statistics/statistics.py]: Changed to if, if since otherwise we have to raise error

File: mlfinlab/backtest_statistics/statistics.py [hunk 20]
```diff
@@ -0,0 +1,268 @@
...
+def all_bets_concentration(returns: pd.Series, frequency: str = 'M') -> tuple:
+    """
+    Snippet 14.3, page 201, Derives positive, negative and time consentration of
+    bets in a given dp.Series of returns.
+    Properties:
+    - low positive_concentration -> no right fat-tail (desirable)
+    - low negative_concentration -> no left fat-tail (desirable)
+    - low time_concentration -> bets are not concentrated in time (desirable)
+    - positive_concentration = 0 ⇔ uniform returns
+    - positive_concentration = 1 ⇔ only one non-zero return
```
[the remark this answers is not in the document]
Comment (PanPip) [reply mlfinlab/backtest_statistics/statistics.py]: Wrote more descriptions

File: mlfinlab/backtest_statistics/statistics.py [hunk 21]
```diff
@@ -0,0 +1,268 @@
...
+    positive_concentration = bets_concentration(returns[returns >= 0])
+    # Concentration of negative returns per bet
+    negative_concentration = bets_concentration(returns[returns < 0])
+    # Concentration of bets/time period (month by default)
+    time_concentration = \
+        bets_concentration(returns.groupby(pd.Grouper(freq=frequency)).count())
+    return (positive_concentration, negative_concentration, time_concentration)
+
+
+def compute_drawdown_and_time_under_water(returns: pd.Series,
```
[the remark this answers is not in the document]
Comment (PanPip) [reply mlfinlab/backtest_statistics/statistics.py]: Will do

File: mlfinlab/tests/test_backtest_statistics.py (hunk 14 above)
Comment (PanPip) [reply mlfinlab/tests/test_backtest_statistics.py]: I haven't seen such explicit tests for function in other files. I would like to know how deep the testing should go?

Comment (Jackal08) [reply mlfinlab/tests/test_backtest_statistics.py]: The tests @Aditya wrote for the portfolio optimisation module sets the best example. 

For several of our modules we wrote tests that arent complete and I want to make sure we correct that mistake going forward.

Comment (aditya1702) [reply mlfinlab/tests/test_backtest_statistics.py]: @Jackal08 Wrong Aditya mentioned. Its @aditya1702 😅

Comment (aditya1702) [reply mlfinlab/tests/test_backtest_statistics.py]: @PanPip You can find the tests in the file test_portfolio_optimisation.py

Comment (PanPip) [reply mlfinlab/tests/test_backtest_statistics.py]: @aditya1702 I can't see explicit tests for NaNs in that file

Comment (aditya1702) [reply mlfinlab/tests/test_backtest_statistics.py]: @PanPip Ah, yes I have not written tests for NaNs. I meant being more stringent in coverage of your code through the tests. For example if there are parts of code which require an error to be thrown then you need to write tests to check if an error is thrown properly or not.

However,  gave your tests a look and they do seem good to me 👍

Comment (PanPip) [reply mlfinlab/tests/test_backtest_statistics.py]: @aditya1702  Thank you :)
